src/leaderboard/leaderboard.service.ts

```diff
@@ -22,47 +20,66 @@ export class LeaderboardService {
-      const list = await this.playerLeaderboardRepository.find({
-        where: {
-          leaderboardId: leaderboardId,
-        },
-      });
-    }
+  async sortLeaderboard(leaderboardId: string): Promise<any> {
+    let temp = 0;
+    let start = 0;
+    let end = 0;
+    let duplicatesIndex = 0;
+    let preSorted = [];
+    let postSorted = [];
 
-    if (sortingOrder[index].includes(SortingOrders.ASC)) {
-      list.sort((x, y) => (x.score[metrics[index]] > y.score[metrics[index]] ? -1 : 1));
-    }
-    if (sortingOrder[index].includes(SortingOrders.DESC)) {
-      list.sort((x, y) => (x.score[metrics[index]] > y.score[metrics[index]] ? 1 : -1));
-    }
+    const leaderboard: Leaderboard = await this.leaderboardRepository.findOne(leaderboardId);
+    let list = await this.playerLeaderboardRepository.find({
+      where: {
+        leaderboardId: leaderboardId,
+      },
+    });
 
-    const tempScores = [];
+    for (let i = 0; i < leaderboard.metrics.length; i++) {
+      let outcome = 1;
+      const metric = leaderboard.metrics[i];
+      const sortingOrder = leaderboard.sortingOrders[i];
+      const reverse = sortingOrder === SortingOrders.DESC ? -1 : 1;
 
-    for (const player of list) {
-      tempScores.push(player.score[metrics[index]]);
-    }
+      list.sort((a, b) => {
+        if (a.score[metric] < b.score[metric]) {
+          return reverse * -1;
+        } else if (a.score[metric] > b.score[metric]) {
+          return reverse * 1;
+        } else {
+          temp = a.score[metric];
+          duplicatesIndex++;
+          outcome = 0;
+          return outcome;
+        }
+      });
 
-    if (this.hasDuplicates(tempScores)) {
-      return this.sortLeaderboard(leaderboardId, leaderboardName, metrics, sortingOrder, index + 1, list);
-    } else {
-      const result: SortedResult = {
-        name: leaderboardName,
-        entries: list,
-      };
-      return result;
-    }
-  }
+      if (preSorted.length !== 0 && postSorted.length === 0) {
+        list.push(...preSorted.concat(list));
+        preSorted = [];
+      } else if (postSorted.length !== 0 && preSorted.length === 0) {
+        list.push(...list.concat(postSorted));
+        postSorted = [];
+      } else if (postSorted.length !== 0 && preSorted.length !== 0) {
+        list = preSorted.concat(list).concat(postSorted);
+        postSorted = [];
+        preSorted = [];
+      }
 
-  hasDuplicates<T>(arr: T[]): boolean {
-    return new Set(arr).size < arr.length;
+      if (duplicatesIndex === 0) {
+        break;
+      }
+
+      if (duplicatesIndex !== 0) {
+        const tempIndex = list.reverse().findIndex(element => element.score[metric] === temp);
+        start = list.reverse().findIndex(element => element.score[metric] === temp);
+        preSorted = list.slice(0, start);
+        end = list.length - tempIndex;
+        postSorted = list.slice(end);
+        duplicatesIndex = 0;
+        list = list.slice(start, end);
+      }
```

Comment:
Here I'm basically finding where the range of players with the same score is placed - `start` and `end`. I've tried using .lastIndexOf() to find the ending but it was producing an error and it appears that it works differently than indexOf - findIndex takes a lambda that returns boolean, while lastindexof takes the object to search for. 

In this part we trim the original array to contain only the range of those players, create a subarray of entries placed before and after this range. Also resetting duplicatesIndex to 0, so as to have a neutral value for the next iteration.